File: advent-of-code/advent_of_code_2016/day_8/part_1_and_2.js
function swipeCard() {
  let screen = blankScreen();
  let commands = parseInput();

  commands.forEach((command) => {
    if (/rect/.test(command)) {
      screen = rectangle(screen, command);
    } else if (/row/.test(command)) {
      screen = rotateRow(screen, command);
    } else if (/column/.test(command)) {
      screen = rotateColumn(screen, command);
    }
  });

  console.log(pixelCount(screen));
  display(screen);
}

function blankScreen() {
  let screen = [];

  for (let i = 0; i < 6; i += 1) {
    screen[i] = new Array(50).fill(' ');
  }

  return screen;
}

function parseInput() {
  function fileRead(filename) {
    let contents = fs.readFileSync(filename);
    return contents.toString();
  }

  let fs = require('fs');
  return fileRead('input.txt').split('\n');
}

function parseCommand(command) {
  return command.match(/\d+/g).map(Number);
}

function rectangle(screen, command) {
  let width, height;
  [width, height] = parseCommand(command);

  for (let i = 0; i < height; i += 1) {
    for (let j = 0; j < width; j += 1) {
      screen[i][j] = '#';
    }
  }

  return screen;
}

function rotateRow(screen, command) {
  let row, shift;
  [row, shift] = parseCommand(command);

  screen[row] = rotate(screen[row], shift);

  return screen;
}

function rotateColumn(screen, command) {
  let col, shift;
  [col, shift] = parseCommand(command);

  screen = transpose(screen);
  screen[col] = rotate(screen[col], shift);

  return transpose(screen);
}

function rotate(array, shift) {
  let copy = [...array];

  return copy.map((el, idx) => {
    let newIndex = idx - shift;
    if (newIndex < 0) newIndex += array.length;
    return array[newIndex];
  });
}

function transpose(matrix) {
  let transposed = [];
  let length = matrix[0].length

  for (let i = 0; i < length; i += 1) {
    transposed[i] = [];
    matrix.forEach((row) => transposed[i].push(row[i]));
  }

  return transposed;
}

function pixelCount(screen) {
  return screen.map((row) => {
    let count = 0;

    row.forEach((px) => {
      if (px === '#') count += 1;
    });

    return count;
  }).reduce((a, b) => a + b);
}

function display(screen) {
  screen.forEach((row) => console.log(row.join('')));
}

swipeCard();

File: advent-of-code/advent_of_code_2016/day_8/part_1_and_2.rb
# start state:
#  - a screen 50px wide and 6px tall
#  - all pixels are off

# 3 types of command:
#  - rect AxB => turns on all the pixels in the top left of the screen
#  - rotate row A by B => shift all of the pixels in a given row (with wrapping)
#  - rotate column A by B => shift all of the pixels in a given column (with wrapping)

# input: list of commands (to be parsed)
# output: number of pixels lit on the screen

class Device
  def initialize(input)
    @commands = input.split("\n")
    @screen = Array.new(6) { Array.new(50, ' ') }
    swipe_card
  end

  def pixel_count
    screen.map { |row| row.count('#') }.sum
  end

  def display_screen
    screen.each { |row| p row.join }
  end

  private

  attr_accessor :screen

  def swipe_card
    @commands.each do |command|
      if command.match?('rect')
        rectangle(command)
      elsif command.match?('row')
        rotate_row(command)
      elsif command.match?('column')
        rotate_column(command)
      end
    end
  end

  def parse(command)
    command.scan(/\d+/).map(&:to_i)
  end

  def rectangle(command)
    width, height = parse(command)

    (0...height).each do |x|
      (0...width).each do |y|
        screen[x][y] = '#'
      end
    end
  end

  def rotate_row(command)
    row, shift = parse(command)
    screen[row].rotate!(-shift)
  end

  def rotate_column(command)
    col, shift = parse(command)

    modified_screen = screen.transpose
    modified_screen[col].rotate!(-shift)

    self.screen = modified_screen.transpose
  end
end

input = File.read('input.txt')
device = Device.new(input)
p device.pixel_count # => 115
device.display_screen # => EFEYKFRFIJ
